src/OmniSharp.Roslyn.CSharp/Services/Structure/BlockStructureService.cs

```diff
@@ -0,0 +1,92 @@
+        private readonly Lazy<Type> _blockStructure;
+        private readonly Lazy<Type> _blockSpan;
+        private readonly Lazy<MethodInfo> _getBlockStructure;
+        private readonly MethodInfo _getSpans;
+        private readonly MethodInfo _getIsCollpasible;
+        private readonly MethodInfo _getTextSpan;
+        private readonly MethodInfo _getHintSpan;
+        private readonly MethodInfo _getBannerText;
+        private readonly MethodInfo _getType;
+        private readonly OmniSharpWorkspace _workspace;
+
+        [ImportingConstructor]
+        public BlockStructureService(IAssemblyLoader loader, OmniSharpWorkspace workspace)
+        {
+            _workspace = workspace;
+            _loader = loader;
+            _featureAssembly = _loader.LazyLoad(Configuration.RoslynFeatures);
+
+            _blockStructureService = _featureAssembly.LazyGetType("Microsoft.CodeAnalysis.Structure.BlockStructureService");
+            _blockStructure = _featureAssembly.LazyGetType("Microsoft.CodeAnalysis.Structure.BlockStructure");
+            _blockSpan = _featureAssembly.LazyGetType("Microsoft.CodeAnalysis.Structure.BlockSpan");
+
+            _getBlockStructure = _blockStructureService.LazyGetMethod("GetBlockStructure");
+            _getSpans = _blockStructure.Value.GetProperty("Spans").GetMethod;
+            _getIsCollpasible = _blockSpan.Value.GetProperty("IsCollapsible").GetMethod;
+            _getTextSpan = _blockSpan.Value.GetProperty("TextSpan").GetMethod;
+            _getHintSpan = _blockSpan.Value.GetProperty("HintSpan").GetMethod;
+            _getBannerText = _blockSpan.Value.GetProperty("BannerText").GetMethod;
+            _getType = _blockSpan.Value.GetProperty("Type").GetMethod;
+        }
+
+        public async Task<BlockStructureResponse> Handle(BlockStructureRequest request)
+        {
+            var document = _workspace.GetDocument(request.FileName);
+            var text = await document.GetTextAsync();
+            var lines = text.Lines;
+
+            var service = _blockStructureService.LazyGetMethod("GetService").InvokeStatic(new[] { document });
+
+            var structure = _getBlockStructure.Invoke<object>(service, new object[] { document, CancellationToken.None });
+            var spans = _getSpans.Invoke<IEnumerable>(structure, Array.Empty<object>());
+
+
+            var outliningSpans = new List<BlockSpan>();
+            foreach (var span in spans)
+            {
+                if (_getIsCollpasible.Invoke<bool>(span, Array.Empty<object>()))
+                {
+                    var textSpan = _getTextSpan.Invoke<TextSpan>(span, Array.Empty<object>());
+                    var line = lines.GetLineFromPosition(textSpan.Start);
+                    var column = textSpan.Start - line.Start;
+                    var endLine = lines.GetLineFromPosition(textSpan.End);
+                    var endColumn = textSpan.End - endLine.Start;
+
+                    outliningSpans.Add(new BlockSpan(
```

Comment:
If I recall correctly, Roslyn's block structure API handles spans for code folding *and* vertical lines. Do we want to include all of the inner spans that are intended for vertical lines in the editor? Or, should this be filtered to just the block spans that are used for outlining?